File: src/main/java/ordilov/randomplay/playlist/domain/PlaylistMapper.java
package ordilov.randomplay.playlist.domain;

import java.util.List;
import ordilov.randomplay.playlist.domain.PlaylistInfo.Item;
import ordilov.randomplay.playlist.domain.PlaylistInfo.PlaylistWithLike;
import org.mapstruct.Mapper;
import org.mapstruct.Mapping;
import org.mapstruct.Mappings;

@Mapper(componentModel = "spring")
public interface PlaylistMapper {

  @Mappings({
      @Mapping(expression = "java(playlist.getPlaylistItems().stream().map(p -> of(p)).collect(java.util.stream.Collectors.toList()))", target = "items")
  })
  PlaylistInfo.Main of(Playlist playlist);

  PlaylistInfo.Main of(PlaylistWithLike playlistWithLike);

  @Mappings({
      @Mapping(expression = "java(item.getTrack().getId())", target = "trackId"),
      @Mapping(expression = "java(item.getTrack().getTitle())", target = "title"),
      @Mapping(expression = "java(item.getTrack().getResourceId())", target = "resourceId"),
  })
  Item of(PlaylistItem item);

  List<PlaylistInfo.Main> of(List<PlaylistWithLike> playlists);
}

File: src/main/java/ordilov/lissn/playlist/domain/PlaylistMapper.kt
package ordilov.lissn.playlist.domain

import org.springframework.stereotype.Component

@Component
class PlaylistMapper {

    fun of(playlist: Playlist): Main {
        return Main(
            id = playlist.id!!,
            title = playlist.title,
            items = playlist.playlistItems.map { of(it) },
            isLiked = false
        )
    }

    fun of(playlistWithLike: PlaylistWithLike): Main {
        return Main(
            id = playlistWithLike.id,
            title = playlistWithLike.title,
            items = playlistWithLike.items,
            isLiked = playlistWithLike.isLiked
        )
    }

    fun of(item: PlaylistItem): Item {
        return Item(
            id = item.id!!,
            trackId = item.track.id,
            resourceId = item.track.resourceId,
            title = item.track.title,
            isLiked = false
        )
    }

    fun of(playlists: List<PlaylistWithLike>): List<Main> {
        return playlists.map { of(it) }
    }
}